ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java

```diff
@@ -110,6 +111,10 @@ boolean receive(LogAppender logAppender, AppendEntriesReplyProto proto,
       return isCompletedNormally();
     }
 
+    boolean isAcknowledged(RaftPeerId id) {
+      return Optional.ofNullable(replies.get(id)).map(HeartbeatAck::isAcknowledged).isPresent();
```

Comment:
What if `replies.get(id)` is not null, but `HeartbeatAck::isAcknowledged` returns false?

```java
      return Optional.ofNullable(replies.get(id)).map(HeartbeatAck::isAcknowledged).getOrElse(false);
```
```java
      return Optional.ofNullable(replies.get(id)).filter(HeartbeatAck::isAcknowledged).isPresent();
```

Reply:
Thanks for helping review the pull request! The current changes have introduced new problems and your two changes are correct. Let me revert this and then submit a new pull request!